### ISSUE TITLE
TrainerEvaluationLoopMixin activates model.train() at the end

### ISSUE THREAD

**Issue by moi90** [PyTorchLightning/pytorch-lightning#2551]
## 🐛 Bug

According to the [example on fine-tuning](https://github.com/jbschiratti/pytorch-lightning/blob/master/pl_examples/domain_templates/computer_vision_fine_tuning.py), it is important to set the frozen sub-modules to eval mode. This is sensitive because when in training mode, BatchNorm and Dropout change state.

However, at the end of `TrainerEvaluationLoopMixin._evaluate` there is following code:

```python
# enable train mode again
model.train()
```

So after the first validation run, the model is again completely in training mode and the freezing is partially undone (for layers like BatchNorm and Dropout).

**Comment by awaelchli** [comment]
I agree, this should not happen in the validation loop. Only the training loop should switch the model to training mode.

**Comment by williamFalcon** [comment]
so, we have to track the state of all the frozen modules before?

**Comment by moi90** [comment]
@awaelchli When accepting that the training mode could have been customized, not even the training loop should change it carelessly.

@williamFalcon That would be one solution.

Another would be to just advice people to set the mode in `on_epoch_start`. (Would this be late enough so it is not reset by the `Trainer`?)

**Comment by awaelchli** [comment]
@williamFalcon @moi90 what if we just made this a hook? The default hook will implement what we have today, set the model to train or eval mode. But if user wants the fine tuning use case, they can override the hook and just set their layers to train/eval mode manually.

**Comment by moi90** [comment]
Where and when would this hook be called? How would this be different from using `on_epoch_start`?

**Comment by awaelchli** [comment]
> Where and when would this hook be called?

Wherever we call model.eval / model.train today in the training loop, we would call the hook instead, which by default also just does that, unless user overrides it.

> How would this be different from using on_epoch_start?

I think this would be different from an on epoch start because it would allow you to prevent exactly what happens as described in the title: activating model.train for all layers at the end of an epoch.

**Comment by edenlightning** [comment]
@awaelchli can this be added post v1?

**Comment by awaelchli** [comment]
Yes

**Comment by moi90** [comment]
Thank you, this is much appreciated!

**Comment by emirkonuk** [comment]
Shouldn't the hooks be defined like:

```
def on_validation_model_eval_end(self) -> None:
  """
  Called in the validation loop after the model is set to eval
  """
  # do something when the model is set to eval
```

How do we use these hooks in the current version? If I don't want to switch to eval mode ever (a pretty common case with GANs), do I just override

```
class DummyCallback(pl.Callback):
  def on_validation_model_eval():
    pass
```

or do I call on_validation_model_train() and then the eval hook won't be called?

**Comment by mhubii** [comment]
Hi, I am looking at an example, where I have kind of a feature extractor as part of the model, hence, would it be possible to also add `on_train_model_train()`, similar to `on_validation_model_train()` or `on_test_model_train` etc.

**Comment by mhubii** [comment]
Hi, I am looking at an example, where I have kind of a feature extractor as part of the model, hence, would it be possible to also add `on_train_model_train`, similar to `on_validation_model_train` and `on_test_model_train` etc.

Could you also call a hook for that? For example here in `Trainer.train()` https://github.com/PyTorchLightning/pytorch-lightning/blob/9d871228ec948ceed0838efdb08953cf458c778d/pytorch_lightning/trainer/trainer.py#L477

**Comment by awaelchli** [comment]
not sure, open an feature request? then people can discuss
